pkg/aws/sts/cache.go

```diff
@@ -77,70 +81,81 @@ func DefaultCache(
 	return c
 }
 
-func (c *credentialsCache) evicted(role string, item interface{}) {
+func (c *credentialsCache) evicted(key string, item interface{}) {
 	f := item.(*future.Future)
 	obj, err := f.Get(context.Background())
 
 	if err != nil {
-		log.WithField("pod.iam.role", role).Debugf("evicted credentials future had error: %s", err.Error())
+		log.WithField("cache.key", key).Debugf("evicted credentials future had error: %s", err.Error())
 		return
 	}
 
-	creds := obj.(*Credentials)
+	cachedCreds := obj.(*CachedCredentials)
 	select {
-	case c.expiring <- &RoleCredentials{Role: role, Credentials: creds}:
-		log.WithFields(CredentialsFields(creds, role)).Infof("notified credentials expire soon")
+	case c.expiring <- cachedCreds:
+		log.WithFields(CredentialsFields(cachedCreds.Identity, cachedCreds.Credentials)).Infof("notified credentials expire soon")
 		return
 	default:
 		return
 	}
 }
 
-func (c *credentialsCache) Expiring() chan *RoleCredentials {
+func (c *credentialsCache) Expiring() chan *CachedCredentials {
 	return c.expiring
 }
 
-func (c *credentialsCache) CredentialsForRole(ctx context.Context, role string) (*Credentials, error) {
-	logger := log.WithFields(log.Fields{"pod.iam.role": role})
-	item, found := c.cache.Get(role)
+func (c *credentialsCache) CredentialsForRole(ctx context.Context, identity *CredentialsIdentity) (*Credentials, error) {
+	logger := log.WithFields(log.Fields{"pod.iam.role": identity.Role})
+	item, found := c.cache.Get(identity.String())
 
 	if found {
 		future, _ := item.(*future.Future)
 		val, err := future.Get(ctx)
 
 		if err != nil {
 			logger.Errorf("error retrieving credentials in cache from future: %s. will delete", err.Error())
-			c.cache.Delete(role)
+			c.cache.Delete(identity.Role)
```

Comment:
`identity.Role` will need to be `identity.String()` to match when it's set and retrieved.

Reply:
Ouch my bad! I guess I had relied on tests a little much here, thanks for the careful eye. This should be fixed.

Reply:
No worries. Yeah, absolutely should have had some tests to catch, sorry.

Reply:
I can look at what coverage I can add as I get through the rest of the work.